electron/execution.js

```diff
@@ -190,34 +208,43 @@ async function onTest(data) {
     }
     /**
      * Fork the Synthetics CLI with correct browser path and
-     * cwd correctly spanws the process
+     * cwd correctly spawns the process
      */
-    const { stdout, stdin, stderr } = fork(`${SYNTHETICS_CLI}`, args, {
+    synthCliProcess = fork(`${SYNTHETICS_CLI}`, args, {
       env: {
         PLAYWRIGHT_BROWSERS_PATH,
       },
       cwd: isDev ? process.cwd() : process.resourcesPath,
       stdio: "pipe",
     });
+    const { stdout, stdin, stderr } = synthCliProcess;
     if (!isSuite) {
       stdin.write(data.code);
       stdin.end();
     }
     stdout.setEncoding("utf-8");
     stderr.setEncoding("utf-8");
     for await (const chunk of stdout) {
-      constructResult(chunk);
+      emitResult(chunk);
     }
     for await (const chunk of stderr) {
       logger.error(chunk);
     }
     if (isSuite) {
       await rm(filePath, { recursive: true, force: true });
     }
-    return result;
-  } catch (err) {
-    logger.error(err);
-    return result;
+  } catch (error) {
+    logger.error(error);
+    sendTestEvent({
+      event: "journey/end",
+      data: {
+        error,
+      },
+    });
+  } finally {
+    if (synthCliProcess) {
+      synthCliProcess.kill();
+    }
```

Comment:
I think a better course of action might be to disable the ability to start recording while testing the previously-recorded journey. There's a testing run status flag internally in the test hook that we could expose, and disallow recording when that value is truthy.

Reply:
Kind of https://github.com/elastic/synthetics-recorder/pull/139 in reverse.

Reply:
This is quite an interesting find, Thanks @lucasfcosta!
Along with what @justinkambic suggested, the backend also needs to ensure running a certain command at a time and abort other requests when needed. Right now all the function calls from the frontend can run simultaneously as they are emitted from individual listeners and don't know about others(whether the journey is running so that you can not record a new one). It makes me think we would need dedicated services per domain, such as a service responsible for communicating with browser/synthetics-cli and a service responsible for application-specific activity(saving a script or link to external) so it is easier to maintain the behaviour in a certain context.

Reply:
I've implemented a UI-only fix for this here: https://github.com/elastic/synthetics-recorder/pull/143/commits/80fe8e3036e6e687e644487cfd604f7e22f4bef3

We'll probably want to address API-level changes in a follow-up as well.

Reply:
created a ticket regarding the API-level change: https://github.com/elastic/synthetics-recorder/issues/149

Reply:
Yes @lucasfcosta, @kyungeunni and I had some offline discussion and came to the same conclusion. The API protection would just be to bake this constraint into the underlying codebase to prevent devs from introducing a regression later.